### PR TITLE
Remove unnecessary calls to `#unsafe_as(UInt64)` etc.

### PR DESCRIPTION
**Pull request by straight-shoota**
This rewrites some use cases of `unsafe_as` casting between integer types. These conversions are not unsafe and don't need `unsafe_as`. They can be expressed with appropriate `to_uX!` calls.

The different overloads of `Hasher#reduce_num` could probably be merged to one or two defs. I'll leave that to a follow-up.

### CODE REVIEW

File: src/crystal/hasher.cr [hunk 1]
```diff
@@ -118,7 +118,7 @@ struct Crystal::Hasher
   end
 
   def self.reduce_num(value : Int)
-    value.remainder(HASH_MODULUS).to_i64.unsafe_as(UInt64)
+    value.remainder(HASH_MODULUS).to_u64!
   end
```
Comment (ysbaddaden) [inline src/crystal/hasher.cr:122]: **nitpick**: this is slightly different: in the very odd case that `value` is an Int128 then the conversion might overflow and this used to raise, as will the above overload with Int::Unsigned :grin:

Comment (straight-shoota) [reply src/crystal/hasher.cr:122]: The result of `remainder(HASH_MODULUS)` is in the range `0...HASH_MODULUS`, and thus guaranteed to fit into `UInt64` (`HASH_MODULUS` is `Int64`).

Comment (HertzDevil) [reply src/crystal/hasher.cr:124]: That would be `#%`, the range for `#remainder` is `-HASH_MODULUS + 1..HASH_MODULUS - 1`

**Review by ysbaddaden — APPROVED** [review]

File: src/crystal/hasher.cr (hunk 1 above)
Comment (straight-shoota) [reply src/crystal/hasher.cr:124]: Oh right. Doesn't change the fact that the result fits in 64-bits and no overflow when casting to `UInt64`.

**Pull request merged**